parl/algorithms/torch/ppo.py

```diff
@@ -17,83 +17,188 @@
+            Args:
+                model (parl.Model): forward network of actor and critic.
+                clip_coef (float): epsilon in clipping loss.
+                vf_coef (float): value function loss coefficient in the optimization objective.
+                ent_coef (float): policy entropy coefficient in the optimization objective.
+                start_lr (float): learning rate.
+                eps (float): Adam optimizer epsilon.
+                max_grad_norm (float): max gradient norm for gradient clipping.
+                clip_vloss (bool): whether or not to use a clipped loss for the value function
+                norm_adv (bool): whether or not to use advantages normalization
+        """
         # checks
         check_model_method(model, 'value', self.__class__.__name__)
         check_model_method(model, 'policy', self.__class__.__name__)
 
-        self.model = model
-
-        self.clip_param = clip_param
-
-        self.value_loss_coef = value_loss_coef
-        self.entropy_coef = entropy_coef
-
+        assert isinstance(clip_coef, float)
+        assert isinstance(vf_coef, float)
+        assert isinstance(ent_coef, float)
+        assert isinstance(start_lr, float)
+        self.start_lr = start_lr
+        self.eps = eps
+        self.clip_coef = clip_coef
+        self.ent_coef = ent_coef
+        self.vf_coef = vf_coef
         self.max_grad_norm = max_grad_norm
-        self.use_clipped_value_loss = use_clipped_value_loss
-
-        self.optimizer = optim.Adam(model.parameters(), lr=initial_lr, eps=eps)
-
-    def learn(self, obs_batch, actions_batch, value_preds_batch, return_batch,
-              old_action_log_probs_batch, adv_targ):
-        values = self.model.value(obs_batch)
-        mean, log_std = self.model.policy(obs_batch)
-        dist = Normal(mean, log_std.exp())
-
-        action_log_probs = dist.log_prob(actions_batch).sum(-1, keepdim=True)
-        dist_entropy = dist.entropy().sum(-1).mean()
-
-        ratio = torch.exp(action_log_probs - old_action_log_probs_batch)
-        surr1 = ratio * adv_targ
-        surr2 = torch.clamp(ratio, 1.0 - self.clip_param,
-                            1.0 + self.clip_param) * adv_targ
-        action_loss = -torch.min(surr1, surr2).mean()
-
-        if self.use_clipped_value_loss:
-            value_pred_clipped = value_preds_batch + \
-                (values - value_preds_batch).clamp(-self.clip_param, self.clip_param)
-            value_losses = (values - return_batch).pow(2)
-            value_losses_clipped = (value_pred_clipped - return_batch).pow(2)
-            value_loss = 0.5 * torch.max(value_losses,
-                                         value_losses_clipped).mean()
+        self.clip_vloss = clip_vloss
+
+        self.norm_adv = norm_adv
+
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        self.model = model.to(device)
+        self.continuous_action = self.model.continuous_action
```

Comment:
It‘s not recommended to get `continue_action` from model, for the model is user-defined